Flaky test: org.keycloak.testsuite.federation.sync.SyncFederationTest#test03ConcurrentSync

## org.keycloak.testsuite.federation.sync.SyncFederationTest#test03ConcurrentSync
[Keycloak CI - Base IT (5)](https://github.com/keycloak/keycloak/actions/runs/4321308439/jobs/7542484711) / [Pull Request #17378](https://github.com/keycloak/keycloak/pull/17378)

```
javax.ws.rs.InternalServerErrorException: HTTP 500 Internal Server Error
	at org.jboss.resteasy.client.jaxrs.internal.ClientInvocation.handleErrorStatus(ClientInvocation.java:278)
	at org.jboss.resteasy.client.jaxrs.internal.ClientInvocation.extractResult(ClientInvocation.java:240)
	at org.jboss.resteasy.client.jaxrs.internal.proxy.extractors.BodyEntityExtractor.extractEntity(BodyEntityExtractor.java:64)
	at org.jboss.resteasy.client.jaxrs.internal.proxy.ClientInvoker.invokeSync(ClientInvoker.java:154)
...
```

I cannot make it fail in my env, even using one cpu only. But my feeling is that the provider cannot be removed because the other transaction in the sync is still doing things. For example:

```
Caused by: org.h2.jdbc.JdbcSQLTransactionRollbackException: Deadlock detected. The current transaction was rolled back. Details: "COMPONENT"; SQL statement:
2023-03-03T09:21:31.4998678Z SyncFederationTest ++ delete from COMPONENT where ID=? [40001-214]
...
2023-03-03T09:21:31.5006943Z SyncFederationTest ++ Caused by: org.h2.mvstore.MVStoreException: Transaction 1 has been chosen as a deadlock victim. Details:
2023-03-03T09:21:31.5007451Z SyncFederationTest ++ Transaction 1 attempts to update map <table.93> entry with key <476> modified by transaction 2(1286) OPEN 31
2023-03-03T09:21:31.5007939Z SyncFederationTest ++ Transaction 2 attempts to update map <table.417> entry with key <2732> modified by transaction 1(1288) OPEN 25
```

So I'm going to modify a bit the test, and the exception in the delete will be swallowed. The next test is going to use another provider name to avoid issues if it cannot be removed.

--- a/testsuite/integration-arquillian/servers/auth-server/services/testsuite-providers/src/main/java/org/keycloak/testsuite/federation/sync/SyncDummyUserFederationProviderFactory.java
+++ b/testsuite/integration-arquillian/servers/auth-server/services/testsuite-providers/src/main/java/org/keycloak/testsuite/federation/sync/SyncDummyUserFederationProviderFactory.java
@@ -42,12 +42,14 @@ import java.util.concurrent.TimeUnit;
 public class SyncDummyUserFederationProviderFactory extends DummyUserFederationProviderFactory {
 
     // Used during SyncFederationTest
-    public static volatile CountDownLatch latch1 = new CountDownLatch(1);
-    public static volatile CountDownLatch latch2 = new CountDownLatch(1);
+    public static volatile CountDownLatch latchStarted = new CountDownLatch(1);
+    public static volatile CountDownLatch latchWait = new CountDownLatch(1);
+    public static volatile CountDownLatch latchFinished = new CountDownLatch(1);
 
     public static void restartLatches() {
-        latch1 = new CountDownLatch(1);
-        latch2 = new CountDownLatch(1);
+        latchStarted = new CountDownLatch(1);
+        latchWait = new CountDownLatch(1);
+        latchFinished = new CountDownLatch(1);
     }
 
 
@@ -77,6 +79,12 @@ public class SyncDummyUserFederationProviderFactory extends DummyUserFederationP
 
     @Override
     public SynchronizationResult syncSince(Date lastSync, KeycloakSessionFactory sessionFactory, String realmId, UserStorageProviderModel model) {
+        if (latchStarted.getCount() <= 0) {
+            logger.info("Already executed, returning");
+            return SynchronizationResult.empty();
+        }
+        // we are starting => allow the test to continue
+        latchStarted.countDown();
 
         KeycloakModelUtils.runJobInTransaction(sessionFactory, new KeycloakSessionTask() {
 
@@ -104,7 +112,8 @@ public class SyncDummyUserFederationProviderFactory extends DummyUserFederationP
 
 
                 try {
-                    latch1.await(waitTime * 1000, TimeUnit.MILLISECONDS);
+                    // await the test to finish
+                    latchWait.await(waitTime * 1000, TimeUnit.MILLISECONDS);
                 } catch (InterruptedException ie) {
                     Thread.currentThread().interrupt();
                     throw new RuntimeException("Interrupted!", ie);
@@ -115,10 +124,10 @@ public class SyncDummyUserFederationProviderFactory extends DummyUserFederationP
 
         });
 
-        // countDown, so the SyncFederationTest can continue
-        latch2.countDown();
+        // countDown, so the SyncFederationTest can finish
+        latchFinished.countDown();
 
-        return new SynchronizationResult();
+        return SynchronizationResult.empty();
     }
 
 }

--- a/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/federation/sync/SyncFederationTest.java
+++ b/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/federation/sync/SyncFederationTest.java
@@ -279,7 +279,7 @@ public class SyncFederationTest extends AbstractAuthTest {
             model.setFullSyncPeriod(-1);
             model.setChangedSyncPeriod(1);
             model.setLastSync(0);
-            model.getConfig().putSingle(SyncDummyUserFederationProviderFactory.WAIT_TIME, "2000");
+            model.getConfig().putSingle(SyncDummyUserFederationProviderFactory.WAIT_TIME, "20");
             ComponentModel dummyModel = new UserStorageProviderModel(appRealm.addComponentModel(model));
         });
 
@@ -295,18 +295,30 @@ public class SyncFederationTest extends AbstractAuthTest {
             usersSyncManager.bootstrapPeriodic(sessionFactory, session.getProvider(TimerProvider.class));
 
             // Wait and then trigger sync manually. Assert it will be ignored
-            sleep(1800);
-            SynchronizationResult syncResult = usersSyncManager.syncChangedUsers(sessionFactory, appRealm.getId(), dummyModel);
-            Assert.assertTrue(syncResult.isIgnored());
-
-            // Cancel timer
-            usersSyncManager.notifyToRefreshPeriodicSync(session, appRealm, dummyModel, true);
-
-            // Signal to factory to finish waiting
-            SyncDummyUserFederationProviderFactory.latch1.countDown();
-
             try {
-                SyncDummyUserFederationProviderFactory.latch2.await(20000, TimeUnit.MILLISECONDS);
+                SyncDummyUserFederationProviderFactory.latchStarted.await(20000, TimeUnit.MILLISECONDS);
+                SynchronizationResult syncResult = usersSyncManager.syncChangedUsers(sessionFactory, appRealm.getId(), dummyModel);
+                Assert.assertTrue(syncResult.isIgnored());
+
+                // Cancel timer
+                usersSyncManager.notifyToRefreshPeriodicSync(session, appRealm, dummyModel, true);
+
+                // Signal to factory to finish waiting
+                SyncDummyUserFederationProviderFactory.latchWait.countDown();
+
+                // wait the task to be finished
+                SyncDummyUserFederationProviderFactory.latchFinished.await(20000, TimeUnit.MILLISECONDS);
+
+                // This sync is here just to ensure that we have lock (doublecheck that periodic sync, which was possibly triggered before canceling timer is finished too)
+                while (true) {
+                    SynchronizationResult result = usersSyncManager.syncChangedUsers(session.getKeycloakSessionFactory(), appRealm.getId(), dummyModel);
+                    if (result.isIgnored()) {
+                        log.infof("Still waiting for lock before periodic sync is finished: %s", result.toString());
+                        sleep(1000);
+                    } else {
+                        break;
+                    }
+                }
             } catch (Exception e) {
                 throw new RuntimeException(e);
             }

